compiler/x/codegen/OMRMemoryReference.cpp

```diff
@@ -1003,7 +1003,6 @@ OMR::X86::MemoryReference::estimateBinaryLength(TR::CodeGenerator *cg)
 
       case 5:
          displacement = self()->getDisplacement();
-         TR_ASSERT(IS_32BIT_SIGNED(displacement), "64-bit displacement should have been replaced in TR_AMD64MemoryReference::generateBinaryEncoding");
```

Comment:
How is simply removing this assertion sufficient? Note the comments on lines 1022 and 1044 that expect the displacement fits in 4 bytes.

Reply:
These asserts are some of the oldest Testarossa “fossils” and were introduced when AMD64 support was first added.  It is questionable whether these were ever correct and useful because they are in the code that conservatively decides how much code cache space this memory reference requires which must always run **before** the binary encoding phase referenced in the assert message (you have to allocate memory before writing into it).

The memory reference code is not the cleanest and could stand refactoring and redesign (after 25 years of experience with it).  The estimate binary length code won’t return an estimate that is too low for memory references with a displacement > 32-bits.  What ends up happening is that on 64-bit we reserve a register to stash displacements > 32-bits into.  There is a parent binary estimator that calls this one and it accounts for the 64-bit offset load into a register.  During the actual binary encoding, on 64-bit the code to process memory references with >32-bit displacements essentially intercepts the binary encoding of the memory reference, inserts the constant materialization instruction, modifies the memory reference to consume the register it was loaded into, and then re-runs the encoding.

This was kind of hacked together when AMD64 was first implemented with the intent on cleaning it up, but it has stuck around.  Some of the comments in the code should be reviewed for currency as well and modified as necessary.

P.S. : these asserts are compiled out of production builds and only appear in debug builds launched by developers or in other downstream projects.

P.P.S. : lest you be thinking that we are wasting code cache space with these conservative estimates, we don’t.  We usually over-allocate the code cache for a method but immediately give back what wasn’t used.

Reply:
Thank you very much for the detailed explanation.